File: python/packages/autogen-agentchat/src/autogen_agentchat/base/_termination.py
import asyncio
from abc import ABC, abstractmethod
from typing import List, Sequence

from autogen_core import Component, ComponentBase, ComponentModel
from pydantic import BaseModel
from typing_extensions import Self

from ..messages import BaseAgentEvent, BaseChatMessage, StopMessage


class TerminatedException(BaseException): ...


class TerminationCondition(ABC, ComponentBase[BaseModel]):
    """A stateful condition that determines when a conversation should be terminated.

    A termination condition is a callable that takes a sequence of ChatMessage objects
    since the last time the condition was called, and returns a StopMessage if the
    conversation should be terminated, or None otherwise.
    Once a termination condition has been reached, it must be reset before it can be used again.

    Termination conditions can be combined using the AND and OR operators.

    Example:

        .. code-block:: python

            import asyncio
            from autogen_agentchat.conditions import MaxMessageTermination, TextMentionTermination


            async def main() -> None:
                # Terminate the conversation after 10 turns or if the text "TERMINATE" is mentioned.
                cond1 = MaxMessageTermination(10) | TextMentionTermination("TERMINATE")

                # Terminate the conversation after 10 turns and if the text "TERMINATE" is mentioned.
                cond2 = MaxMessageTermination(10) & TextMentionTermination("TERMINATE")

                # ...

                # Reset the termination condition.
                await cond1.reset()
                await cond2.reset()


            asyncio.run(main())
    """

    component_type = "termination"

    @property
    @abstractmethod
    def terminated(self) -> bool:
        """Check if the termination condition has been reached"""
        ...

    @abstractmethod
    async def __call__(self, messages: Sequence[BaseAgentEvent | BaseChatMessage]) -> StopMessage | None:
        """Check if the conversation should be terminated based on the messages received
        since the last time the condition was called.
        Return a StopMessage if the conversation should be terminated, or None otherwise.

        Args:
            messages: The messages received since the last time the condition was called.

        Returns:
            StopMessage | None: A StopMessage if the conversation should be terminated, or None otherwise.

        Raises:
            TerminatedException: If the termination condition has already been reached."""
        ...

    @abstractmethod
    async def reset(self) -> None:
        """Reset the termination condition."""
        ...

    def __and__(self, other: "TerminationCondition") -> "TerminationCondition":
        """Combine two termination conditions with an AND operation."""
        return AndTerminationCondition(self, other)

    def __or__(self, other: "TerminationCondition") -> "TerminationCondition":
        """Combine two termination conditions with an OR operation."""
        return OrTerminationCondition(self, other)


class AndTerminationConditionConfig(BaseModel):
    conditions: List[ComponentModel]


class AndTerminationCondition(TerminationCondition, Component[AndTerminationConditionConfig]):
    component_config_schema = AndTerminationConditionConfig
    component_type = "termination"
    component_provider_override = "autogen_agentchat.base.AndTerminationCondition"

    def __init__(self, *conditions: TerminationCondition) -> None:
        self._conditions = conditions
        self._stop_messages: List[StopMessage] = []

    @property
    def terminated(self) -> bool:
        return all(condition.terminated for condition in self._conditions)

    async def __call__(self, messages: Sequence[BaseAgentEvent | BaseChatMessage]) -> StopMessage | None:
        if self.terminated:
            raise TerminatedException("Termination condition has already been reached.")
        # Check all remaining conditions.
        stop_messages = await asyncio.gather(
            *[condition(messages) for condition in self._conditions if not condition.terminated]
        )
        # Collect stop messages.
        for stop_message in stop_messages:
            if stop_message is not None:
                self._stop_messages.append(stop_message)
        if any(stop_message is None for stop_message in stop_messages):
            # If any remaining condition has not reached termination, it is not terminated.
            return None
        content = ", ".join(stop_message.content for stop_message in self._stop_messages)
        source = ", ".join(stop_message.source for stop_message in self._stop_messages)
        return StopMessage(content=content, source=source)

    async def reset(self) -> None:
        for condition in self._conditions:
            await condition.reset()
        self._stop_messages.clear()

    def _to_config(self) -> AndTerminationConditionConfig:
        """Convert the AND termination condition to a config."""
        return AndTerminationConditionConfig(conditions=[condition.dump_component() for condition in self._conditions])

    @classmethod
    def _from_config(cls, config: AndTerminationConditionConfig) -> Self:
        """Create an AND termination condition from a config."""
        conditions = [TerminationCondition.load_component(condition_model) for condition_model in config.conditions]
        return cls(*conditions)


class OrTerminationConditionConfig(BaseModel):
    conditions: List[ComponentModel]
    """List of termination conditions where any one being satisfied is sufficient."""


class OrTerminationCondition(TerminationCondition, Component[OrTerminationConditionConfig]):
    component_config_schema = OrTerminationConditionConfig
    component_type = "termination"
    component_provider_override = "autogen_agentchat.base.OrTerminationCondition"

    def __init__(self, *conditions: TerminationCondition) -> None:
        self._conditions = conditions

    @property
    def terminated(self) -> bool:
        return any(condition.terminated for condition in self._conditions)

    async def __call__(self, messages: Sequence[BaseAgentEvent | BaseChatMessage]) -> StopMessage | None:
        if self.terminated:
            raise RuntimeError("Termination condition has already been reached")
        stop_messages = await asyncio.gather(*[condition(messages) for condition in self._conditions])
        if any(stop_message is not None for stop_message in stop_messages):
            content = ", ".join(stop_message.content for stop_message in stop_messages if stop_message is not None)
            source = ", ".join(stop_message.source for stop_message in stop_messages if stop_message is not None)
            return StopMessage(content=content, source=source)
        return None

    async def reset(self) -> None:
        for condition in self._conditions:
            await condition.reset()

    def _to_config(self) -> OrTerminationConditionConfig:
        """Convert the OR termination condition to a config."""
        return OrTerminationConditionConfig(conditions=[condition.dump_component() for condition in self._conditions])

    @classmethod
    def _from_config(cls, config: OrTerminationConditionConfig) -> Self:
        """Create an OR termination condition from a config."""
        conditions = [TerminationCondition.load_component(condition_model) for condition_model in config.conditions]
        return cls(*conditions)

File: dotnet/src/Microsoft.AutoGen/AgentChat/Abstractions/Termination.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Termination.cs

namespace Microsoft.AutoGen.AgentChat.Abstractions;

public static class TerminationConditionExtensions
{
    /// <summary>
    /// Combine this termination condition with another using a logical OR.
    /// </summary>
    /// <param name="other">Another termination condition.</param>
    /// <returns>The combined termination condition, with appropriate short-circuiting.</returns>
    public static ITerminationCondition Or(this ITerminationCondition this_, ITerminationCondition other)
    {
        return new CombinerCondition(CombinerCondition.Or, this_, other);
    }

    /// <summary>
    /// Combine this termination condition with another using a logical AND.
    /// </summary>
    /// <param name="other">Another termination condition.</param>
    /// <returns>The combined termination condition, with appropriate short-circuiting.</returns>
    public static ITerminationCondition And(this ITerminationCondition this_, ITerminationCondition other)
    {
        return new CombinerCondition(CombinerCondition.And, this_, other);
    }
}

/// <summary>
/// A stateful condition that determines when a conversation should be terminated.
///
/// A termination condition takes a sequences of <see cref="AgentMessage"/> objects since the last time the
/// condition was checked, and returns a <see cref="StopMessage"/> if the conversation should be terminated,
/// or <c>null</c> otherwise.
///
/// Once a termination condition has been reached, it must be <see cref="Reset()"/> before it can be used again.
///
/// Termination conditions can be combined using the <see cref="TerminationConditionExtensions.Or"/> and
/// <see cref="TerminationConditionExtensions.And"/> methods.
/// </summary>
public interface ITerminationCondition
{
    /// <summary>
    /// Checks if the termination condition has been reached
    /// </summary>
    public bool IsTerminated { get; }

    /// <summary>
    /// Check if the conversation should be terminated based on the messages received
    /// since the last time the condition was called.
    /// Return a <see cref="StopMessage"/> if the conversation should be terminated, or <c>null</c> otherwise.
    /// </summary>
    /// <param name="messages">The messages received since the last time the condition was called.</param>
    /// <returns>A <see cref="StopMessage"/> if the conversation should be terminated, or <c>null</c>
    /// otherwise.</returns>
    /// <exception cref="TerminatedException">If the termination condition has already been reached.</exception>
    public ValueTask<StopMessage?> CheckAndUpdateAsync(IList<AgentMessage> messages);

    /// <summary>
    /// Resets the termination condition.
    /// </summary>
    public void Reset();

    /// <summary>
    /// Combine two termination conditions with another using an associative, short-circuiting OR.
    /// </summary>
    /// <param name="left">
    /// The left-hand side termination condition. If this condition is already a disjunction, the RHS condition is added to the list of clauses.
    /// </param>
    /// <param name="right">
    /// The right-hand side termination condition. If the LHS condition is already a disjunction, this condition is added to the list of clauses.
    /// </param>
    /// <returns>
    /// The combined termination condition, with appropriate short-circuiting.
    /// </returns>
    public static ITerminationCondition operator |(ITerminationCondition left, ITerminationCondition right)
    {
        return left.Or(right);
    }

    /// <summary>
    /// Combine two termination conditions with another using an associative, short-circuiting AND.
    /// </summary>
    /// <param name="left">
    /// The left-hand side termination condition. If this condition is already a conjunction, the RHS condition is added to the list of clauses.
    /// </param>
    /// <param name="right">
    /// The right-hand side termination condition. If the LHS condition is already a conjunction, this condition is added to the list of clauses.
    /// </param>
    /// <returns>
    /// The combined termination condition, with appropriate short-circuiting.
    /// </returns>
    public static ITerminationCondition operator &(ITerminationCondition left, ITerminationCondition right)
    {
        return left.And(right);
    }
}

/// <summary>
/// Exception thrown when a termination condition has already been reached.
/// </summary>
public sealed class TerminatedException : Exception
{
    public TerminatedException() : base("The termination condition has already been reached.")
    {
    }
}

/// <summary>
/// A termination condition that combines multiple termination conditions using a logical AND or OR.
/// </summary>
internal sealed class CombinerCondition : ITerminationCondition
{
    public const bool Conjunction = true;
    public const bool Disjunction = false;

    public const bool And = Conjunction;
    public const bool Or = Disjunction;

    private List<StopMessage> stopMessages = new List<StopMessage>();
    private List<ITerminationCondition> clauses;
    private readonly bool conjunction;

    /// <summary>
    /// Create a new <see cref="CombinerCondition"/> with the given conjunction and clauses.
    /// </summary>
    /// <param name="conjunction">The conjunction to use when combining the clauses.</param>
    /// <param name="clauses">The termination conditions to combine.</param>
    public CombinerCondition(bool conjunction, params IEnumerable<ITerminationCondition> clauses)
    {
        // Flatten the list of clauses by unwrapping included CombinerConditions if their
        // conjunctions match (since combiners with associative conjunctions can be hoisted).
        IEnumerable<ITerminationCondition> flattened =
            clauses.SelectMany(c =>
                    (c is CombinerCondition combiner && combiner.conjunction == conjunction)
                    ? (IEnumerable<ITerminationCondition>)combiner.clauses
                    : new[] { c });

        this.conjunction = conjunction;

        this.clauses = flattened.ToList();
    }

    /// <inheritdoc cref="ITerminationCondition.IsTerminated" />
    public bool IsTerminated { get; private set; }

    /// <inheritdoc cref="ITerminationCondition.Reset" />
    public void Reset()
    {
        this.stopMessages.Clear();
        this.clauses.ForEach(c => c.Reset());

        this.IsTerminated = false;
    }

    /// <inheritdoc cref="ITerminationCondition.CheckAndUpdateAsync" />
    public async ValueTask<StopMessage?> CheckAndUpdateAsync(IList<AgentMessage> messages)
    {
        if (this.IsTerminated)
        {
            throw new TerminatedException();
        }

        // When operating as a conjunction, we may be accumulated terminating conditions, but we will not fire until
        // all of them are complete. In this case, we need to avoid continuing to interact with already terminated
        // clauses, because trying to update them will throw
        var candidateTerminations = this.conjunction ? this.clauses.Where(clause => !clause.IsTerminated) : clauses;

        // TODO: Do we really need these to be ValueTasks? (Alternatively: Do we really need to run them explicitly
        // on every invocation, or is a Worker pattern more appropriate?)
        List<Task<StopMessage?>> tasks = candidateTerminations.Select(c => c.CheckAndUpdateAsync(messages).AsTask()).ToList();
        StopMessage?[] results = await Task.WhenAll(tasks);

        bool raiseTermination = this.conjunction; // if or, we start with false until we observe a true
                                                  // if and, we start with true until we observe a false

        foreach (StopMessage? maybeStop in results)
        {
            if (maybeStop != null)
            {
                this.stopMessages.Add(maybeStop);
                if (!this.conjunction)
                {
                    // If any clause terminates, the disjunction terminates
                    raiseTermination = true;
                }
            }
            else if (this.conjunction)
            {
                // If any clause does not terminate, the conjunction does not terminate
                raiseTermination = false;
            }
        }

        if (raiseTermination)
        {
            this.IsTerminated = true;

            return new StopMessage
            {
                Content = string.Join("; ", stopMessages.Select(sm => sm.Content)),
                Source = string.Join(", ", stopMessages.Select(sm => sm.Source))
            };
        }

        return null;
    }
}
